### ISSUE TITLE
RSPEC-138 | JS | Functions should not have too many lines of code | React Functional Component

### ISSUE THREAD

**Issue by VovaSv** [SonarSource/SonarJS#1762]
Description
We have this issue in our projects on each React Functional Component which can be more than 200 line of code.

Repro steps
Please provide the steps required to reproduce the problem

Run this rule of any React Functional Component which more than 200 line of code
Expected behavior
Ignore for React Functional Component

Actual behavior
Applied on each React Function Component

**Comment by vilchik-elena** [comment]
Hi @VovaSv,

Could you provide version of analyzer where you have this problem?
Also could you explain why React Functional Component is allowed to be >200?
To help us to improve please provide detailed explanation of the pattern you think should be excluded by the rule, in other words how we can know if some function is React Functional Component. 

Also do you see any other React patterns which should be ignored by this rule?

**Comment by VovaSv** [comment]
Hi Elena,

First big thanks for your quick response.

Regarding your questions:
1. We have use in our organization sonarqube version:
     SonarQube™ technology is powered by SonarSource SA
<http://www.sonarsource.com/>
- Community Edition
- Version 7.5 (build 20543).
-
-
-
-
-
-
-
-

2. There is no limitation for React Functional component length. React
components based on JSX. JSX stands for JavaScript XML. JSX allows us to
write HTML elements in JavaScript.
     So when we write function component it can include JS with HTML
together which easy can passed 200 lines of code.
     Also, React Component rise another issue in sonar related to
Refactor(*Critical
called Cognitive Complexity of functions should not be too high*).
     Example: Refactor this function to reduce its Cognitive Complexity
from 46 to the 15 allowed.

*I think we need add some annotation (like //REACTCOMPONENT  like we have
for //NOSONAR) which will indicate to sonar for ignore in specific rules.*



Please see below example:
To many lines:
[image: image.png]


Refactor:
[image: image.png]



*Thanks & Regards*,

Vova Svidinsky

**Comment by yananym** [comment]
Concur with the statement above. Basically React functional component has it's own methods and can easily grow beyond 200 lines. Think of it as a class - would you but a limit of 200 lines on a JS class?

**Comment by yanivnahoum** [comment]
I agree that React Functional Components should be treated as classes - [they are an alternative syntax for Class Components](https://reactjs.org/docs/components-and-props.html):
```javascript
function Welcome(props) {
  return <h1>Hello, {props.name}</h1>;
}
```

```javascript
class Welcome extends React.Component {
  render() {
    return <h1>Hello, {this.props.name}</h1>;
  }
}
```
The above two components are equivalent from React’s point of view. It makes sense then for sonarlint to treat them as such, and compute the cognitive complexity of Component Functions with the same algorithm used for classes.

**Comment by guillaume-dequenne-sonarsource** [comment]
Related community feedback for cognitive complexity: https://community.sonarsource.com/t/sonarqube-counts-incorrect-cognitive-complexity-for-react-functional-components/30183

**Comment by Vladqqqqoo** [comment]
I got an answer. You can take a part in this discussion.
https://community.sonarsource.com/t/sonarqube-counts-incorrect-cognitive-complexity-for-react-functional-components-bug-no-functionality/30852/2
